packages/arb-validator-core/gobridge/messagesChallenge.go

```diff
@@ -0,0 +1,391 @@
+	ethMsg, _ := value.NewTupleFromSlice([]value.Value{
+		value.NewIntValue(msg.BlockNum.AsInt()),
+		value.NewIntValue(msg.MessageNum),
+		msgType,
+	})
+	ethMsgHash := ethMsg.Hash()
+
+	msgs, _ := value.NewTupleFromSlice([]value.Value{
+		value.NewHashOnlyValue(lowerHashB, 32),
+		value.NewHashOnlyValue(ethMsgHash, 32),
+	})
+	matchHash := valprotocol.MessageChallengeDataHash(
+		lowerHashA,
+		hashing.SoliditySHA3(
+			hashing.Bytes32(lowerHashA),
+			hashing.Bytes32(messageHash)),
+		lowerHashB,
+		msgs.Hash(),
+		big.NewInt(1),
+	)
+
+	if !c.client.GoEthClient.challenges[c.contractAddress].challengerDataHash.Equals(matchHash) {
+		return errors.New("OneStepProofERC20Message Incorrect previous state")
+	}
+
+	c.client.GoEthClient.pubMsg(c.challengeData, arbbridge.MaybeEvent{
+		Event: arbbridge.OneStepProofEvent{
+			ChainInfo: arbbridge.ChainInfo{
+				BlockId: c.client.GoEthClient.getCurrentBlock(),
+			},
+		},
+	})
+	// TODO: handle stake distribution
+	//	_asserterWin();
+	//		resolveChallengeAsserterWon();
+	//			require(challenges[msg.sender], RES_CHAL_SENDER);
+	//			delete challenges[msg.sender];
+	//
+	//			Staker storage winningStaker = getValidStaker(address(winner));
+	//			winner.transfer(stakeRequirement / 2);
+	//			winningStaker.inChallenge = false;
+	//			deleteStaker(loser);
+	//
+	//			emit RollupChallengeCompleted(msg.sender, address(winner), loser);
+	c.client.GoEthClient.pubMsg(c.challengeData, arbbridge.MaybeEvent{
+		Event: arbbridge.ChallengeCompletedEvent{
+			ChainInfo: arbbridge.ChainInfo{
+				BlockId: c.client.GoEthClient.getCurrentBlock(),
+			},
+			Winner:            msg.From,
+			Loser:             msg.To,
+			ChallengeContract: c.contractAddress,
+		},
+	})
+	return nil
+}
+
+func (c *messagesChallenge) OneStepProofERC721Message(
+	ctx context.Context,
+	lowerHashA common.Hash,
+	lowerHashB common.Hash,
+	msg message.DeliveredERC721,
+) error {
+	messageHash := msg.CommitmentHash()
+
+	msgType := msg.AsValue()
+	ethMsg, _ := value.NewTupleFromSlice([]value.Value{
+		value.NewIntValue(msg.BlockNum.AsInt()),
+		value.NewIntValue(msg.MessageNum),
+		msgType,
+	})
+	ethMsgHash := ethMsg.Hash()
+
+	msgs, _ := value.NewTupleFromSlice([]value.Value{
+		value.NewHashOnlyValue(lowerHashB, 32),
+		value.NewHashOnlyValue(ethMsgHash, 32),
+	})
+	matchHash := valprotocol.MessageChallengeDataHash(
+		lowerHashA,
+		hashing.SoliditySHA3(
+			hashing.Bytes32(lowerHashA),
+			hashing.Bytes32(messageHash)),
+		lowerHashB,
+		msgs.Hash(),
+		big.NewInt(1),
+	)
+
+	if !c.client.GoEthClient.challenges[c.contractAddress].challengerDataHash.Equals(matchHash) {
+		return errors.New("OneStepProofERC721Message Incorrect previous state")
+	}
+
+	c.client.GoEthClient.pubMsg(c.challengeData, arbbridge.MaybeEvent{
+		Event: arbbridge.OneStepProofEvent{
+			ChainInfo: arbbridge.ChainInfo{
+				BlockId: c.client.GoEthClient.getCurrentBlock(),
+			},
+		},
+	})
+	// TODO: handle stake distribution
+	//	_asserterWin();
```

Comment:
The challenge currently doesn't seem to call the rollup chain to tell it the challenge is resolved.

Reply:
implemented resolveChallenge